README.md
# Cylon.js For Raspberry Pi

Cylon.js (http://cylonjs.com) is a JavaScript framework for robotics, physical computing, and the Internet of Things (IoT).

This module provides an adaptor for the Raspberry Pi single board computer (http://www.raspberrypi.org/). It currently works with all models of Raspberry Pi including the Raspberry Pi 2 and Raspberry Pi 3.

Want to use Ruby on robots? Check out our sister project Artoo (http://artoo.io)

Want to use the Go programming language to power your robots? Check out our sister project Gobot (http://gobot.io).

[![Build Status](https://secure.travis-ci.org/hybridgroup/cylon-raspi.png?branch=master)](http://travis-ci.org/hybridgroup/cylon-raspi) [![Code Climate](https://codeclimate.com/github/hybridgroup/cylon-raspi/badges/gpa.svg)](https://codeclimate.com/github/hybridgroup/cylon-raspi) [![Test Coverage](https://codeclimate.com/github/hybridgroup/cylon-raspi/badges/coverage.svg)](https://codeclimate.com/github/hybridgroup/cylon-raspi)

## How to Install

Installing Cylon.js for the Raspberry Pi is easy, but must be done on the Raspi itself, or on another Linux computer. Due to I2C device support, the module cannot be installed on OS X or Windows.

Install the module with:

    $ npm install cylon cylon-raspi cylon-gpio cylon-i2c

## How to Use

This small program causes an LED to blink.

```javascript
var Cylon = require("cylon");

Cylon.robot({
  connections: {
    raspi: { adaptor: 'raspi' }
  },

  devices: {
    led: { driver: 'led', pin: 11 }
  },

  work: function(my) {
    every((1).second(), my.led.toggle);
  }
}).start();
```

## How to Connect

### Install the latest Raspbian OS

You can get it from here: [http://www.raspberrypi.org/downloads/](http://www.raspberrypi.org/downloads/)

### Setting the Raspberry Pi keyboard

Having trouble with your Raspberry Pi keyboard layout? Use the following command:

    sudo dpkg-reconfigure keyboard-configuration

### Update your Raspbian and install Node.js

These commands need to be run after SSHing into the Raspi:

    sudo apt-get update
    sudo apt-get upgrade

### Install Node.js

You can either install the latest Node.js via the Nodesource Debian repo, or else use `nvm` (https://github.com/creationix/nvm).

To use the Debian package, first add the Nodesource package repo:

    curl -sL https://deb.nodesource.com/setup_5.x | sudo -E bash -

Then install the latest Node.js by running this command:

    sudo apt-get install -y nodejs

You should see the node version you just installed.

    $ node --version
    v5.6.0

### Connecting to Raspberry Pi GPIO

This module only works on a real Raspberry Pi. Do not bother trying on any other kind of computer, it will not work.
Also note you will need to connect actual circuits to the Raspberry Pi's GPIO pins.

In order to access the GPIO pins without using `sudo` you will need to both app the pi user to the `gpio` group:

    sudo usermod -G gpio pi

And also add the following udev rules file to `/etc/udev/rules.d/91-gpio.rules`:

    SUBSYSTEM=="gpio", KERNEL=="gpiochip*", ACTION=="add", PROGRAM="/bin/sh -c 'chown root:gpio /sys/class/gpio/export /sys/class/gpio/unexport ; chmod 220 /sys/class/gpio/export /sys/class/gpio/unexport'"
    SUBSYSTEM=="gpio", KERNEL=="gpio*", ACTION=="add", PROGRAM="/bin/sh -c 'chown root:gpio /sys%p/active_low /sys%p/direction /sys%p/edge /sys%p/value ; chmod 660 /sys%p/active_low /sys%p/direction /sys%p/edge /sys%p/value'"

Thanks to "MikeDK" for the above solution: [https://www.raspberrypi.org/forums/viewtopic.php?p=198148#p198148](https://www.raspberrypi.org/forums/viewtopic.php?p=198148#p198148)

### Enabling the Raspberry Pi i2c on Raspbian

#### Easy Install

The really easy way to configure your i2c is to use the `i2c-tools` extensions for `raspi-config`. You can install it like this:

```
sudo apt-get install i2c-tools
```

Once installed, run `sudo raspi-config` and choose the "Advanced Settings" menu. Select the "I2C Settings" and follow the prompts. You will need to reboot after changing the settings.

#### Manual Install

You must add these two entries to your `/etc/modules`

    i2c-bcm2708
    i2c-dev

You must also ensure that these entries are commented in your `/etc/modprobe.d/raspi-blacklist.conf`

    #blacklist spi-bcm2708
    #blacklist i2c-bcm2708

You will also need to update the /boot/config.txt file. Edit it add the following text:

    dtparam=i2c1=on
    dtparam=i2c_arm=on

Finally, you need to allow the `pi` user permissions to access the i2c interface by running this command:

    sudo usermod -G i2c pi

Now restart your Raspberry Pi.

### Enabling PWM output on GPIO pins.

You need to install and have pi-blaster running in the raspberry-pi, you can follow the instructions for pi-blaster install in the pi-blaster repo here:

[https://github.com/sarfata/pi-blaster](https://github.com/sarfata/pi-blaster)

### Available PINS

The following object depicts available pins for all revisions of raspberry-pi, the `key` is the actual number of the physical pin header on the board,the `value` is the GPIO pin number assigned by the OS, for the pins with changes between board revisions, the value contains the variations of GPIO pin number assignment between them (eg.rev1, rev2, rev3).

You should just be concerned with the `key` (number of the physical pin header on the board), Cylon.JS takes care of the board revision and GPIO pin numbers for you, this full list is for reference only.

```javascript
PINS = {
  3: {
    rev1: 0,
    rev2: 2,
    rev3: 2
  },
  5: {
    rev1: 1,
    rev2: 3,
    rev3: 3
  },
  7: 4,
  8: 14,
  10: 15,
  11: 17,
  12: 18,
  13: {
    rev1: 21,
    rev2: 27,
    rev3: 27
  },
  15: 22,
  16: 23,
  18: 24,
  19: 10,
  21: 9,
  22: 25,
  23: 11,
  24: 8,
  29: {
    rev3: 5
  },
  31: {
    rev3: 6
  },
  32: {
    rev3: 12
  },
  33: {
    rev3: 13
  },
  35: {
    rev3: 19
  },
  36: {
    rev3: 16
  },
  37: {
    rev3: 26
  },
  38: {
    rev3: 20
  },
  40: {
    rev3: 21
  }
};
```

The website http://pi.gadgetoid.com/pinout has a great visual representation of this information.

## Documentation

We're busy adding documentation to our web site at http://cylonjs.com/ please check there as we continue to work on Cylon.js

Thank you!

## Contributing

For our contribution guidelines, please go to [https://github.com/hybridgroup/cylon/blob/master/CONTRIBUTING.md
](https://github.com/hybridgroup/cylon/blob/master/CONTRIBUTING.md
).

## Release History

For the release history, please go to [https://github.com/hybridgroup/cylon-raspi/blob/master/RELEASES.md
](https://github.com/hybridgroup/cylon-raspi/blob/master/RELEASES.md
).

## License
Copyright (c) 2013-2016 The Hybrid Group. Licensed under the Apache 2.0 license.


RELEASES.md
## Release History

Version 0.20.1 - Correct problem with i2c block read params

Version 0.20.0 - Use Cylon 1.3.0, cleanup dependencies

Version 0.19.0 - Rewrite to use i2c-bus module, now supports all current versions of Node.js

Version 0.18.0 - Use Cylon 1.2.0

Version 0.17.0 - Use Cylon 1.1.0

Version 0.16.0 - Compatibility with Cylon 1.0.0

Version 0.15.3 - Corrected functionality for i2c.

Version 0.15.2 - Remove node-i2c dependency from package, it is now install via script

Version 0.15.1 - Install node-i2c using script based on node version

Version 0.15.0 - Compatibility with Cylon 0.22.0

Version 0.14.1 - Fix pin and pi-blaster compatability for different board revisions

Version 0.14.0 - Compatibility with Cylon 0.21.0

Version 0.13.0 - Compatibility with Cylon 0.20.0

Version 0.12.0 - Compatibility with Cylon 0.19.0

Version 0.11.0 - Compatibility with Cylon 0.18.0

Version 0.10.0 - Compatibility with Cylon 0.16.0

Version 0.9.1 - Add peerDependencies to package.json

Version 0.9.0 - Compatibility with Cylon 0.15.0

Version 0.8.1 - Fix a bug with PWM pins

Version 0.8.0 - Compatibility with Cylon 0.14.0, remove node-namespace.

Version 0.7.0 - Release for cylon 0.12.0

Version 0.6.2 - Fixes bug with connection not being set on adaptor.

Version 0.6.0 - Release for cylon 0.11.0, refactor into pure JavaScript, i2c bugfixes

Version 0.5.0 - Release for cylon 0.10.0

Version 0.4.0 - Release for cylon 0.9.0

Version 0.3.0 - Release for cylon 0.8.0

Version 0.2.0 - Release for cylon 0.7.0

Version 0.1.0 - Initial release


lib/raspi.js
/*
 * Cylonjs Raspi adaptor
 * http://cylonjs.com
 *
 * Copyright (c) 2013-2016 The Hybrid Group
 * Licensed under the Apache 2.0 license.
*/

"use strict";

var Cylon = require("cylon");

var PwmPin = require("./pwm-pin"),
    I2CDevice = require("./i2c-device"),
    fs = require("fs");

var PINS = {
  3: {
    rev1: 0,
    rev2: 2,
    rev3: 2
  },
  5: {
    rev1: 1,
    rev2: 3,
    rev3: 3
  },
  7: 4,
  8: 14,
  10: 15,
  11: 17,
  12: 18,
  13: {
    rev1: 21,
    rev2: 27,
    rev3: 27
  },
  15: 22,
  16: 23,
  18: 24,
  19: 10,
  21: 9,
  22: 25,
  23: 11,
  24: 8,
  26: 7,
  29: { rev3: 5 },
  31: { rev3: 6 },
  32: { rev3: 12 },
  33: { rev3: 13 },
  35: { rev3: 19 },
  36: { rev3: 16 },
  37: { rev3: 26 },
  38: { rev3: 20 },
  40: { rev3: 21 }
};

var Raspi = module.exports = function Raspi() {
  Raspi.__super__.constructor.apply(this, arguments);

  this.board = "";
  this.pins = {};
  this.pwmPins = {};
  this.i2cDevices = {};
};

Cylon.Utils.subclass(Raspi, Cylon.Adaptor);

Raspi.prototype.commands = [
  "pins",
  "pinMode",

  "firmwareName",

  "digitalRead",
  "digitalWrite",

  "pwmWrite",
  "servoWrite",

  "i2cWrite",
  "i2cRead"
];

/**
 * Connects to the Raspberry Pi
 *
 * @param {Function} callback to be triggered when connected
 * @return {void}
 */
Raspi.prototype.connect = function(callback) {
  this.proxyMethods(this.commands, this.board, this);

  var cpuinfo = this._cpuinfo();
  var revisionCode = cpuinfo.match(/Revision\s*:\s*([\da-fA-F]+)/)[1];
  var revision = parseInt(revisionCode, 16);

  this.bus = 1;

  if (revision <= 3) {
    this.revision = "rev1";
    this.bus = 0;
  } else if (revision <= 15) {
    this.revision = "rev2";
  } else {
    this.revision = "rev3";
  }

  Cylon.Logger.debug("Raspberry Pi ", this.revision, " detected.");
  callback();
};

/**
 * Disconnects from the Raspberry Pi
 *
 * @param {Function} callback to be triggered when disconnected
 * @return {void}
 */
Raspi.prototype.disconnect = function(callback) {
  Cylon.Logger.debug("Disconnecting all pins...");
  this._disconnectPins();
  Cylon.Logger.debug("Disconnecting from board '" + this.name + "'...");
  this.emit("disconnect");
  callback();
};

Raspi.prototype.firmwareName = function() {
  return "Raspberry Pi";
};

/**
 * Reads a value from a digital pin
 *
 * @param {Number} pinNum pin to read from
 * @param {Function} callback triggered when the value has been read from the
 * pin
 * @return {void}
 * @publish
 */
Raspi.prototype.digitalRead = function(pinNum, callback) {
  var pin = this.pins[this._translatePin(pinNum)];

  if (pin == null) {
    pin = this._digitalPin(pinNum, "r");

    pin.on("digitalRead", function(val) {
      this.respond("digitalRead", callback, null, val, pinNum);
    }.bind(this));

    // We listen for the connect event to make sure the pin
    // has been setup in Linux IO first. Once connected we
    // trigger the digitalRead, this happens only once.
    pin.on("connect", function() {
      pin.digitalRead(20);
    });

    pin.connect();
  }

  return true;
};

/**
 * Writes a value to a digital pin
 *
 * @param {Number} pinNum pin to write to
 * @param {Number} value value to write to the pin
 * @param {Function} callback function to invoke when done
 * @return {void}
 * @publish
 */
Raspi.prototype.digitalWrite = function(pinNum, value, callback) {
  var pin = this.pins[this._translatePin(pinNum)];

  if (pin != null) {
    pin.digitalWrite(value);
  } else {
    pin = this._digitalPin(pinNum, "w");

    pin.on("digitalWrite", function(val) {
      this.respond("digitalWrite", callback, null, val, pinNum);
    }.bind(this));

    // We listen for the connect event to make sure the pin
    // has been setup in Linux IO first. Once connected we
    // trigger the digitalWrite, this happens only once.
    pin.on("connect", function() {
      pin.digitalWrite(value);
    });

    pin.connect();
  }

  return value;
};

/**
 * Writes an I2C value to the board.
 *
 * @param {Number} address I2C address to write to
 * @param {Number} cmd I2C command to write
 * @param {Array} buff buffered data to write
 * @param {Function} callback function to call when done
 * @return {void}
 * @publish
 */
Raspi.prototype.i2cWrite = function(address, cmd, buff, callback) {
  buff = buff || [];

  this._i2cDevice(address).write(cmd, buff, function() {
    this.respond("i2cWrite", callback, null, address, cmd, buff);
  }.bind(this));
};

/**
 * Reads an I2C value from the board.
 *
 * @param {Number} address I2C address to write to
 * @param {Number} cmd I2C command to write
 * @param {Number} length amount of data to read
 * @param {Function} callback function to call with data
 * @return {void}
 * @publish
 */
Raspi.prototype.i2cRead = function(address, cmd, length, callback) {
  this._i2cDevice(address).read(cmd, length, function(err, data) {
    this.respond("i2cRead", callback, err, data);
  }.bind(this));
};

Raspi.prototype._i2cDevice = function(address) {
  if (this.i2cDevices[address] == null) {
    this.i2cDevices[address] = new I2CDevice({
      address: address,
      bus: this.bus
    });
    this.i2cDevices[address].connect();
  }
  return this.i2cDevices[address];
};

/**
 * Writes a PWM value to a pin
 *
 * @param {Number} pinNum pin to write to
 * @param {Number} value value to write to the pin
 * @param {Number} callback function to call when done
 * @param {Number} type write type
 * @return {void}
 * @private
 */
Raspi.prototype._pwmWrite = function(pinNum, value, callback, type) {
  var pin;
  pin = this._pwmPin(pinNum);

  type = type || "pwm";

  pin.on("pwmWrite", function() {
    this.respond(type + "Write", callback, null, value, pinNum);
  }.bind(this));

  if (type === "servo") {
    pin.servoWrite(value);
  } else {
    pin.pwmWrite(value);
  }
};

/**
 * Writes a pwm value to a pin
 *
 * @param {Number} pinNum pin to write to
 * @param {Number} angle angle to write to the pin
 * @param {Number} callback function to call when done
 * @return {void}
 * @publish
 */
Raspi.prototype.pwmWrite = function(pinNum, angle, callback) {
  this._pwmWrite(pinNum, angle, callback, "pwm");
};

/**
 * Writes a servo value to a pin
 *
 * @param {Number} pinNum pin to write to
 * @param {Number} angle angle to write to the pin
 * @param {Number} callback function to call when done
 * @return {void}
 * @publish
 */
Raspi.prototype.servoWrite = function(pinNum, angle, callback) {
  this._pwmWrite(pinNum, angle, callback, "servo");
};

Raspi.prototype._pwmPin = function(pinNum) {
  var gpioPinNum;
  gpioPinNum = this._translatePin(pinNum);

  if (this.pwmPins[gpioPinNum] == null) {
    this.pwmPins[gpioPinNum] = new PwmPin({
      pin: gpioPinNum
    });
  }

  return this.pwmPins[gpioPinNum];
};

Raspi.prototype._digitalPin = function(pinNum, mode) {
  var gpioPinNum;
  gpioPinNum = this._translatePin(pinNum);

  if (this.pins[gpioPinNum] == null) {
    this.pins[gpioPinNum] = new Cylon.IO.DigitalPin({
      pin: gpioPinNum,
      mode: mode
    });
  }

  return this.pins[gpioPinNum];
};

Raspi.prototype._translatePin = function(pinNum) {
  var pin = PINS[pinNum];

  if (typeof pin === "object") {
    return pin[this.revision];
  }

  return pin;
};

Raspi.prototype._disconnectPins = function() {
  var pin;

  for (pin in this.pins) {
    this.pins[pin].closeSync();
  }

  for (pin in this.pwmPins) {
    this.pwmPins[pin].closeSync();
  }
};

Raspi.prototype._cpuinfo = function() {
  return fs.readFileSync("/proc/cpuinfo", { encoding: "utf-8" });
};
